impl/src/main/java/io/kroxylicious/testing/kafka/invm/ReflectionUtils.java

```diff
@@ -0,0 +1,96 @@
+            Void.TYPE, Void.TYPE);
+
+    @NonNull
+    @SuppressWarnings("unchecked")
+    public static <K> Optional<K> construct(Class<K> clazz, Object... parameters) {
+        Constructor<?>[] declaredConstructors = clazz.getDeclaredConstructors();
+        return Arrays.stream(declaredConstructors)
+                .filter(constructor -> Modifier.isPublic(constructor.getModifiers()))
+                .filter(constructor -> matchingMethod(constructor.getParameterTypes(), parameters))
+                .findFirst().map(constructor -> {
+                    try {
+                        return (K) constructor.newInstance(parameters);
+                    }
+                    catch (InstantiationException | IllegalAccessException | InvocationTargetException e) {
+                        throw new IllegalStateException(e);
+                    }
+                });
+    }
+
+    public static <R> R invokeInstanceMethod(@NonNull Object target, @NonNull String methodName, Object... parameters) {
+        var methods = target.getClass().getMethods();
+        return invokeMethod(target.getClass(), target, methodName, parameters, methods,
+                method -> Modifier.isPublic(method.getModifiers()) && !Modifier.isStatic(method.getModifiers()));
+    }
+
+    public static <R> R invokeStaticMethod(@NonNull Class<?> clazz, @NonNull String methodName, Object... parameters) {
+        var methods = clazz.getDeclaredMethods();
+        return invokeMethod(clazz, null, methodName, parameters, methods,
+                method -> Modifier.isPublic(method.getModifiers()) && Modifier.isStatic(method.getModifiers()));
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <R> R invokeMethod(@NonNull Class<?> clazz, @Nullable Object target, @NonNull String methodName, Object[] parameters, Method[] methods,
+                                      Predicate<Method> methodPredicate) {
+        Optional<Method> first = Arrays.stream(methods)
+                .filter(method -> methodName.equals(method.getName()))
+                .filter(methodPredicate)
+                .filter(method -> matchingMethod(method.getParameterTypes(), parameters))
+                .findFirst();
+
+        var method = first.orElseThrow(() -> new UnsupportedOperationException("Can't find method %s on class %s".formatted(methodName, clazz.getName())));
+        try {
+            return (R) method.invoke(target, parameters);
+        }
+        catch (IllegalAccessException | InvocationTargetException e) {
+            throw new IllegalStateException("Failed to invoke method %s on object %s".formatted(methodName, clazz), e);
+        }
+    }
+
+    private static boolean matchingMethod(Class<?>[] methodParameterTypes, Object... parameters) {
+        if (methodParameterTypes.length != parameters.length) {
+            return false;
+        }
+        boolean allMatch = true;
+        for (int i = 0; i < parameters.length; i++) {
+            // What about nulls?
```

Comment:
It looks like we don't expect to handle nulls anywhere, so maybe hoist the non-null parameter requirement to the top `invoke*` and `construct`.

Or we could change it to return true if it's non-primitive and parameter null. Then in `invokeMethod` collect to List, and throw if size != 1, so we only attempt the invoke if it's unambiguous.

Reply:
I think we should just reject any invocations with null args.   Like you say, we don't need any right now.  If we ever do, we can fix it then.